packages/SwingSet/src/vats/vattp/vat-vattp.js

```diff
@@ -16,181 +23,255 @@ import { Far } from '@endo/marshal';
+  // QUESTION: Is there a convention around capitalization here?
+  const mailboxKindHandle = provideKindHandle(baggage, 'mailbox');
+  const initMailboxState = name => ({
+    name,
+    outboundHighestAdded: 0,
+    outboundHighestRemoved: 0,
+    inboundHighestDelivered: 0,
+    inboundReceiver: null,
+  });
+  const makeMailbox = defineDurableKindMulti(
+    mailboxKindHandle,
+    initMailboxState,
+    {
+      transmitter: {
+        transmit: ({ state }, msg) => {
+          const num = state.outboundHighestAdded + 1;
+          D(mailboxDevice).add(state.name, num, msg);
+          state.outboundHighestAdded = num;
+        },
       },
-      onClose(connection, ..._args) {
-        receivers.delete(connection);
-        console.warn(`deleting connection is not fully supported in comms`);
+      setReceiver: {
+        setReceiver: ({ state }, receiver) => {
+          assert(!state.inboundReceiver, X`setReceiver is call-once`);
+          assert(receiver, X`receiver must not be empty`);
+          state.inboundReceiver = receiver;
+        },
       },
-      infoMessage(...args) {
-        void E(cons).log('VatTP connection info:', ...args);
+      inbound: {
+        deliverMessages: ({ state }, newMessages) => {
+          // Minimize interactions with durable storage,
+          // assuming state won't change underneath us
+          // and remote methods won't synchronously throw.
+          const { name, inboundReceiver, inboundHighestDelivered } = state;
+          let newHighestDelivered = inboundHighestDelivered;
+          for (const [num, body] of newMessages) {
+            if (num > newHighestDelivered) {
+              // TODO: SO() / sendOnly()
+              E(inboundReceiver).receive(body);
+              newHighestDelivered = num;
+              D(mailboxDevice).ackInbound(name, num);
+            }
+          }
+          if (newHighestDelivered > inboundHighestDelivered) {
+            state.inboundHighestDelivered = newHighestDelivered;
+          }
+        },
+        deliverAck: ({ state }, ack) => {
+          // Minimize interactions with durable storage,
+          // assuming state won't change underneath us
+          // and remote methods won't synchronously throw.
+          const { name, outboundHighestAdded, outboundHighestRemoved } = state;
+          let newHighestRemoved = outboundHighestRemoved;
+          while (
+            newHighestRemoved < outboundHighestAdded &&
+            newHighestRemoved < ack
+          ) {
+            newHighestRemoved += 1;
+            D(mailboxDevice).remove(name, newHighestRemoved);
+          }
+          if (newHighestRemoved > outboundHighestRemoved) {
+            state.outboundHighestRemoved = newHighestRemoved;
+          }
+        },
       },
-    });
+    },
+  );
 
-    return harden({ host, handler });
+  /** @type {MapStore<string, ReturnType<typeof makeMailbox>>} */
+  const mailboxes = provideDurableMapStore(baggage, 'mailboxes');
```

Comment:
I'm sympathetic to that pattern, but it seems a bit at odds with `vivifySingleton` (which ensures initialization of a key in baggage) being forced to come late because of the guidance to define all variables before their references.

My current compromise is an early "_Define all durable baggage keys and kind handles_" block and use of its constants to define kinds, collections, and durable values later on with preceding "_define kind_"/"_retain collection_" comments, at the cost of long "…BaggageKey" variable names.

Reply:
Yeah, I'm torn too. @erights added a `/* eslint-disable no-use-before-define */` to the top of vat-timer.js for me, so I consider that permission to ignore that bit of guidance :). But I'm also torn between putting all the durable things and Kind definitions in one place (where a future maintainer can easily see what they're obligated to maintain), and using convenient tools like `vivifySingleton` that don't lend themselves to that approach. See https://github.com/Agoric/agoric-sdk/pull/5862#issuecomment-1201955257 for some discussion.

Any approach you choose is good by me.